fml/methods/linear_regression/linear_regression_function_impl.hpp

```diff
@@ -38,18 +38,62 @@ double LinearRegressionFunction<DatasetType, LabelsType>::Evaluate(E& parameters
   return loss + reg;
 }
 
+template<typename DatasetType, typename LabelsType>
+template<typename E>
+double LinearRegressionFunction<DatasetType, LabelsType>::Evaluate(const E& parameters,
+                                                                   const size_t& firstId,
+                                                                   const size_t& batchSize)
+{
+  const size_t lastId = firstId + batchSize;
+
+  auto error = xt::linalg::dot(xt::view(dataset, xt::range(firstId, lastId), xt::all()),
+      parameters) - xt::view(labels, xt::range(firstId, lastId), xt::all());
+
+  // Evaluate the regularization
+  double reg = (lambda / (2 * numFunctions()))
+      * xt::linalg::dot(xt::transpose(parameters), parameters)();
+
+  // Evaluate loss.
+  double loss = xt::linalg::dot(xt::transpose(error), error
+      / (2 * batchSize))();
+
+  return loss + reg;
+}
+
+
 template<typename DatasetType, typename LabelsType>
 template<typename E, typename G>
-void LinearRegressionFunction<DatasetType, LabelsType>::Gradient(E& parameters,
+void LinearRegressionFunction<DatasetType, LabelsType>::Gradient(const E& parameters,
                                                                  G& gradient)
 {
   // Evaluates the error between the evaluated values and
   // actual values.
   auto error = xt::linalg::dot(dataset, parameters) - labels;
 
   // Evaluate the regularized gradient.
-  gradient = xt::linalg::dot(xt::transpose(dataset), error
-      / numFunctions()) + ((lambda / (2 * numFunctions())) * parameters);
+  gradient = xt::transpose(xt::linalg::dot(xt::transpose(error)
+      / numFunctions(), dataset)) + ((lambda / (2 * numFunctions())) * parameters);
```

Comment:
This should improve the speed of computation by avoiding the cost of transposing the `dataset`.